[AMDGPU][Verifier] Mark calls to entry functions as invalid in the IR verifier

shiltian opened the pull request:
For AMDGPU, calls to entry functions are invalid. Previously, due to certain
limitations, this restriction was not enforced by the IR verifier. These
limitations have now been resolved, enabling us to enforce this check.

arsenm commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -3765,9 +3765,28 @@ void Verifier::visitCallBase(CallBase &Call) {
+  if (Function *F = Call.getCalledFunction()) {
     if (Intrinsic::ID ID = (Intrinsic::ID)F->getIntrinsicID())
       visitIntrinsicCall(ID, Call);
+    // TODO: Move this to target dependent IR verifier once we have it.
```
Remove this comment. This belongs in the generic verifier. The calling conventions, while intended to only be used by certain targets, are not target code. The calling conventions are global invariants

arsenm commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -3765,9 +3765,28 @@ void Verifier::visitCallBase(CallBase &Call) {
     if (Intrinsic::ID ID = (Intrinsic::ID)F->getIntrinsicID())
       visitIntrinsicCall(ID, Call);
+    // TODO: Move this to target dependent IR verifier once we have it.
+    auto IsAMDGPUEntryFunctionCC = [](CallingConv::ID CC) -> bool {
```
Reinventing utility function in backend, can move to calling convention header. And rename to something more generic, like isCallableCC or something

arsenm commented on llvm/test/CodeGen/AMDGPU/call-to-kernel.ll:
```diff
@@ -1,18 +1,105 @@
+
+define amdgpu_ls void @caller_ls(ptr addrspace(1) %out) {
+entry:
+  call void @callee_ls(ptr addrspace(1) %out)
```
Broken callsite with wrong convention. You should also separately test the calling convention on the callsite (e.g indirect call)

lalaniket8 commented on llvm/include/llvm/IR/CallingConv.h:
```diff
@@ -290,6 +290,26 @@ namespace CallingConv {
+  case CallingConv::AMDGPU_LS:
+  case CallingConv::AMDGPU_PS:
+  case CallingConv::AMDGPU_VS:
+    return false;
```
Should we also add `CallingConv::SPIR_KERNEL` here since its not supported as well : https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp#L1137

shiltian replied:
There are "valid" test case calling to SPIR-V kernel directly. I thought that is allowed by the SPIR-V spec?

arsenm replied:
It should not be allowed. Also SPIR_KERNEL actually came from spir classic and SPIR-V just started picking it up.

arsenm replied:
"– No function can be targeted by both an OpEntryPoint instruction and an OpFunctionCall instruction."

MrSidims replied:
I lean to disagree. SPIR_KERNEL _can_ be included _conditionally_. If triple is `spirv`, then sure, validation rule makes sense. But for `spir` triple it does look like a breaking change, as there is no such restriction in OpenCL or SPIR specification.

Naghasan replied:
> But for spir triple it does look like a breaking change, as there is no such restriction in OpenCL or SPIR specification

+1, this is valid in these conditions

arsenm commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -3598,14 +3598,8 @@ void Verifier::visitCallBase(CallBase &Call) {
-        "not allowed. Please use the @llvm.amdgpu.cs.chain intrinsic instead.",
-        Call);
+  // Verify if the calling convention of the callee is callable.
+  Check(isCallableCC(Call.getCallingConv()), "Callee is not callable", Call);
```
```suggestion
  Check(isCallableCC(Call.getCallingConv()), "calling convention is does not permit calls", Call);
```

shiltian replied:
FWIW, the verifier isn't very consistent about whether the first letter of an error message should be capitalized.

arsenm replied:
It's not great, but the verifier errors are less user facing than most things

arsenm commented on llvm/test/Verifier/call-to-non-callable-functions.ll:
```diff
@@ -0,0 +1,181 @@
+entry:
+  call amdgpu_ls void %func()
+  ret void
+}
```
Also check an invoke for good measure

shiltian committed:
[AMDGPU][Verifier] Mark calls to entry functions as invalid in the IR verifier

For AMDGPU, calls to entry functions are invalid. Previously, due to certain
limitations, this restriction was not enforced by the IR verifier. These
limitations have now been resolved, enabling us to enforce this check.

Adding target-dependent checks directly into the IR verifier is not ideal.
However, a cleaner solution, such as a dedicated target-dependent IR verifier,
is underway (e.g., https://github.com/llvm/llvm-project/pull/123609). Once that
or similar code is merged, we can move this check accordingly.

MrSidims reviewed (commented):
This patch breaks tests for non AMDGPU code, see https://github.com/KhronosGroup/SPIRV-LLVM-Translator/actions/runs/14434681816/job/40473808013

From OpenCL POW it's legal for a kernel to call another kernel.

arsenm commented:
> This patch breaks tests for non AMDGPU code, see https://github.com/KhronosGroup/SPIRV-LLVM-Translator/actions/runs/14434681816/job/40473808013
> 
> From OpenCL POW it's legal for a kernel to call another kernel.

Which is entirely a clang problem and now correctly implemented in 642481a4286c9006958274531ee173b347866c50
